File: Pointers/Module.cs
using System.Text;

namespace Pointers;

/// <summary>
/// A class representing a Module, containing a sorted list of Students.
/// </summary>
public class Module
{
    public string ModuleId { get; set; }
    public readonly List<Student> _students = new(5);
    public IComparer<Student> StudentsComparer { get; }

    public Module(string moduleId, IComparer<Student>? studentsComparer = null)
    {
        ModuleId = moduleId;
        
        // If no Student Comparer is selected, use the default comparer (by StudentId).
        studentsComparer ??= new StudentIdComparer();
        StudentsComparer = studentsComparer;
    }

    public IReadOnlyList<Student> Students => _students;
    public int ClassCount => _students.Count;

    /// <summary>
    /// Adds a new specified Student to the Module.
    /// </summary>
    /// <param name="newStudent"></param>
    /// <returns>True, if enrolled.</returns>
    public bool Enroll(Student newStudent)
    {
        if (_students.Contains(newStudent))
            return false;
        
        _students.Add(newStudent);
        _students.Sort(StudentsComparer);
        return true;
    }

    /// <summary>
    /// Removes a specified Student from the Module.
    /// </summary>
    /// <param name="student"></param>
    /// <returns>True, if disenrolled.</returns>
    public bool Disenroll(Student student)
    {
        return _students.Remove(student);
    }

    public override string ToString()
    {
        var sb = new StringBuilder(15 * ClassCount);

        sb.AppendLine($"Class: {ModuleId}");
        
        for (var i = 0; i < _students.Count; i++)
            sb.AppendLine($"Student {i}: {_students[i].ToString()}");

        return sb.ToString();
    }
}

File: cis2344/src/pointers/Module.java
package pointers;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * A class representing a Module, containing a sorted list of Students.
 */
public class Module {
    private String moduleId;
    private List<Student> students;
    private Comparator<Student> studentComparator;

    public Module(String moduleId, Comparator<Student> studentComparator) {
        this.moduleId = moduleId;
        students = new ArrayList<>(5);

        // If no Student Comparator is selected, use the default comparator (by StudentId).
        if (studentComparator == null)
            studentComparator = new StudentIdComparator();

        this.studentComparator = studentComparator;
    }

    public String getModuleId() {
        return moduleId;
    }

    public List<Student> getStudents() {
        return students;
    }

    public int getClassCount() {
        return students.size();
    }

    /**
     * Adds a new specified Student to the Module.
     * @param student The new student to enroll.
     * @return True if enrolled.
     */
    public boolean enroll(Student student) {
        if (student == null)
            return false;

        if (students.contains(student))
            return false;

        students.add(student);
        students.sort(studentComparator);

        return true;
    }

    /**
     * Removes a specified Student from the Module.
     * @param student The student to disenroll.
     * @return True if the student was disenrolled.
     */
    public boolean disenroll(Student student) {
        return students.remove(student);
    }

    @Override
    public String toString() {
        var sb = new StringBuilder(15 * getClassCount());

        sb.append("Class: ");
        sb.append(getModuleId());
        sb.append("\n");

        for (int i = 0; i < students.size(); i++) {
            sb.append("Student ");
            sb.append(i);
            sb.append(": ");
            sb.append(students.get(i).toString());
            sb.append("\n");
        }

        return sb.toString();
    }
}
